src/rdkafka_sasl_oauthbearer.c

```diff
@@ -0,0 +1,1006 @@
+                                     const void *buf, size_t size,
+                                     char *errstr, size_t errstr_size) {
+        const rd_chariov_t in = { .ptr = (char *)buf, .size = size };
+        return rd_kafka_sasl_oauthbearer_fsm(rktrans, &in, errstr, errstr_size);
+}
+
+
+/**
+ * @brief Initialize and start SASL OAUTHBEARER (builtin) authentication.
+ *
+ * Returns 0 on successful init and -1 on error.
+ *
+ * @locality broker thread
+ */
+static int rd_kafka_sasl_oauthbearer_client_new (rd_kafka_transport_t *rktrans,
+                                    const char *hostname,
+                                    char *errstr, size_t errstr_size) {
+        struct rd_kafka_sasl_oauthbearer_state *state;
+
+        state = rd_calloc(1, sizeof(*state));
+        state->state = RD_KAFKA_SASL_OAUTHB_STATE_SEND_CLIENT_FIRST_MESSAGE;
+
+        /*
+         * Save off the state structure now, before any possibility of
+         * returning, so that we will always free up the allocated memory in
+         * rd_kafka_sasl_oauthbearer_close().
+         */
+        rktrans->rktrans_sasl.state = state;
+
+        /*
+         * Make sure we have a consistent view of the token and extensions
+         * throughout the authentication process -- even if it is refreshed
+         * midway through this particular authentication.
+         */
+        rd_kafka_rdlock(rktrans->rktrans_rkb->rkb_rk);
+        if (!rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->token_value) {
+                rd_snprintf(errstr, errstr_size,
+                        "OAUTHBEARER cannot log in because there is no token "
+                        "available; last error: %s",
+                        rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->errstr ?
+                        rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->errstr :
+                        "not available");
+                rd_kafka_rdunlock(rktrans->rktrans_rkb->rkb_rk);
+                return -1;
+        }
+        state->token_value = strdup(
+                rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->token_value);
+        state->md_principal_name = strdup(
+                rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->md_principal_name);
+        rd_list_copy_to(&state->extensions,
+                &rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->extensions,
+                rd_strtup_list_copy, NULL);
+        rd_kafka_rdunlock(rktrans->rktrans_rkb->rkb_rk);
+
+        /* Kick off the FSM */
+        return rd_kafka_sasl_oauthbearer_fsm(rktrans, NULL, errstr, errstr_size);
+}
+
+
+
+/**
+ * @brief Validate OAUTHBEARER config, which is a no-op
+ * (we rely on initial token retrieval)
+ */
+static int rd_kafka_sasl_oauthbearer_conf_validate (rd_kafka_t *rk,
+                                              char *errstr,
+                                              size_t errstr_size) {
+        /*
+         * We will rely on the initial token retrieval as a proxy
```

Comment:
We could validate the unsecure JWS here, if we wanted to.

Reply:
It occurs to me now that we can't do the validation here because the config format is implementation-dependent; it might not be the config format that the default unsecured JWS callback handler knows how to parse.  I think therefore we **must** rely on the initial token retrieval as a proxy for validation.  I've changed the comment to reflect this.  Let me know if you disagree.